File: Data Structures & Algorithms/minimum-interval-including-query/submission-0.py
import heapq

class Solution:
    def minInterval(self, intervals, queries):

        intervals.sort()

        # (query value, original index)
        sortedQueries = sorted(
            [(q, i) for i, q in enumerate(queries)]
        )

        res = [-1] * len(queries)

        minHeap = []

        i = 0

        for q, idx in sortedQueries:

            # add all intervals starting before query
            while i < len(intervals) and intervals[i][0] <= q:

                left, right = intervals[i]

                length = right - left + 1

                heapq.heappush(minHeap, (length, right))

                i += 1

            # remove invalid intervals
            while minHeap and minHeap[0][1] < q:
                heapq.heappop(minHeap)

            # smallest valid interval
            if minHeap:
                res[idx] = minHeap[0][0]

        return res

File: Data Structures & Algorithms/frequency-of-the-most-frequent-element/submission-0.cpp
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k) {

        sort(nums.begin(), nums.end());

        long long sum = 0;
        int left = 0;
        int ans = 1;

        for (int right = 0; right < nums.size(); right++) {

            sum += nums[right];

            long long cost =
                (long long)nums[right] * (right - left + 1) - sum;

            while (cost > k) {

                sum -= nums[left];
                left++;

                cost =
                    (long long)nums[right] * (right - left + 1) - sum;
            }

            ans = max(ans, right - left + 1);
        }

        return ans;
    }
};